### PR TITLE
feat: add rng support

### LINKED ISSUE #62: Add new operator RNG

### PR DESCRIPTION
**Pull request by gabaldon**
Avoid using filters and operators in the source script.
close #62

### CODE REVIEW

File: test/src/AggregationTallyScript.spec.ts [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import { I18n } from '../../src/i18n'
 import { AggregationTallyScript } from '../../src/aggregationTallyScript'
 import { Cache } from '../../src/structures'
-import { AggregationTallyFilter, MirAggregationTallyScript } from '../../src/types'
+import { AggregationTallyFilter, Kind, MirAggregationTallyScript } from '../../src/types'
```
Comment (Tommytrg) [inline test/src/AggregationTallyScript.spec.ts:4]:
```suggestion
import { AggregationTallyFilter, SourceType, MirAggregationTallyScript } from '../../src/types'
```

File: test/src/AggregationTallyScript.spec.ts [hunk 2]
```diff
@@ -12,7 +12,7 @@ describe('AggregationTallyScript', () => {
         reducer: 0x02,
       }
       const context = { cache: new Cache(), i18n: new I18n() }
-      const script = new AggregationTallyScript(context, mirScript)
+      const script = new AggregationTallyScript(context, mirScript, Kind.HttpGet)
```
Comment (Tommytrg) [inline test/src/AggregationTallyScript.spec.ts:15]:
```suggestion
      const script = new AggregationTallyScript(context, mirScript, SourceType.HttpGet)
```

File: src/aggregationTallyScript.ts [hunk 3]
```diff
@@ -13,11 +14,13 @@ export class AggregationTallyScript {
   public mirScript: MirAggregationTallyScript
   public reducer: AggregationTallyOperatorReducer
   public scriptId: number
+  public sourceType: Kind
```
Comment (Tommytrg) [inline src/aggregationTallyScript.ts:17]:
```suggestion
  public kind: Kind
```

File: src/aggregationTallyScript.ts [hunk 4]
```diff
@@ -13,11 +14,13 @@ export class AggregationTallyScript {
...
 
-  constructor(context: Context, script: MirAggregationTallyScript) {
+  constructor(context: Context, script: MirAggregationTallyScript, sourceType: Kind) {
```
Comment (Tommytrg) [inline src/aggregationTallyScript.ts:19]:
```suggestion
  constructor(context: Context, script: MirAggregationTallyScript, sourceType: SourceType) {
```

File: src/aggregationTallyScript.ts [hunk 5]
```diff
@@ -70,6 +73,13 @@ export class AggregationTallyScript {
     }
   }
 
+  public updateSourceType(sourceType: Kind) {
```
Comment (Tommytrg) [inline src/aggregationTallyScript.ts:78]:
```suggestion
  public updateSourceType(sourceType: SourceType) {
```

File: src/aggregationTallyScript.ts [hunk 6]
```diff
@@ -70,6 +73,13 @@ export class AggregationTallyScript {
...
+    this.sourceType = sourceType
+    if (this.sourceType === Kind.RNG) {
```
Comment (Tommytrg) [inline src/aggregationTallyScript.ts:80]:
```suggestion
    if (this.sourceType === SourceType.RNG) {
```

File: src/argument.ts [hunk 7]
```diff
@@ -66,7 +67,7 @@ export class Argument {
         argument as Reducer
       )
     } else if (this.argumentInfo.type === MirArgumentType.Subscript) {
-      this.argument = new Script(this.context, argument as MirScript, OutputType.SubscriptOutput)
+      this.argument = new Script(this.context, argument as MirScript, Kind.HttpGet, OutputType.SubscriptOutput)
```
Comment (Tommytrg) [inline src/argument.ts:70]:
```suggestion
      this.argument = new Script(this.context, argument as MirScript, SourceType.HttpGet, OutputType.SubscriptOutput)
```

File: src/constants.ts [hunk 8]
```diff
@@ -1,5 +1,7 @@
-import { OperatorCode, OutputType } from './types'
+import { OperatorCode, OutputType, Kind } from './types'
```
Comment (Tommytrg) [inline src/constants.ts:1]:
```suggestion
import { OperatorCode, OutputType, SourceType } from './types'
```

File: src/constants.ts [hunk 9]
```diff
@@ -1,5 +1,7 @@
...
 
 export const DEFAULT_OPERATOR = OperatorCode.ArrayCount
 export const DEFAULT_INPUT_TYPE = OutputType.Array
 export const DEFAULT_SCRIPT_FIRST_TYPE = OutputType.String
+export const KIND_OPTIONS = [Kind.HttpGet, Kind.RNG]
```
Comment (Tommytrg) [inline src/constants.ts:6]:
```suggestion
export const KIND_OPTIONS = [SourceTypeHttpGet, SourceType.RNG]
```

File: src/radon.ts [hunk 10]
```diff
@@ -13,15 +14,22 @@ export class Radon {
   public retrieve: Array<Source>
   public aggregate: AggregationTallyScript
   public tally: AggregationTallyScript
-
+  private _sourceType: Kind
+  public get sourceType() {
+    return this._sourceType;
+  }
```
Comment (Tommytrg) [inline src/radon.ts:20]: We can access `.sourceType` directly

```suggestion
```

File: src/radon.ts [hunk 11]
```diff
@@ -13,15 +14,22 @@ export class Radon {
...
+  public set sourceType(kind: Kind) {
+    this._sourceType = kind;
+  }
```
Comment (Tommytrg) [inline src/radon.ts:23]:
```suggestion
  public setKind(kind: Kind) {
    this.sourceType = kind
  }
```

File: src/radon.ts [hunk 12]
```diff
@@ -32,14 +40,29 @@ export class Radon {
     ;(this.context.cache.get(scriptId) as Script).addOperator()
   }
 
+  public onChildrenEvent() {
+    return {
+      emit: (e: { sourceType: Kind }) => {
```
Comment (Tommytrg) [inline src/radon.ts:45]:
```suggestion
      emit: (e: { kind: Kind }) => {
```

File: src/radon.ts [hunk 13]
```diff
@@ -32,14 +40,29 @@ export class Radon {
...
+        this.sourceType = e.sourceType
```
Comment (Tommytrg) [inline src/radon.ts:46]:
```suggestion
        this.kind = e.kind
```

File: src/radon.ts [hunk 14]
```diff
@@ -32,14 +40,29 @@ export class Radon {
...
+        if (this.sourceType === Kind.RNG) {
```
Comment (Tommytrg) [inline src/radon.ts]:
```suggestion
        if (this.kind === Kind.RNG) {
```

**Review by Tommytrg — CHANGES_REQUESTED** [review]
I'd rename the variable `sourceType` with `kind` to keep the code homogeneous and to make easier a future integration with witnet-requests-js

File: src/radon.ts (hunk 14 above)
Comment (gabaldon) [reply src/radon.ts]: It makes sense to name the type of the source as `kind` in the `source.ts` file. In other files the naming should be more explicative. I think, in this case, sourceType is a better name.

File: src/script.ts [hunk 15]
```diff
@@ -48,18 +52,19 @@ export class Script {
...
+        const operator: MirOperator = getDefaultMirOperatorByType(type)
+        this.operators.push(
+          new Operator(this.context, this.scriptId, lastOutputType, operator, this.onChildrenEvent())
+        )
+      } else {
+        // TODO: search in operators the type for the regarding types:
+        // SubscriptOutput, ReducerOutput, FilterOutput, MatchOutput, Same, Inner
+        this.operators.push(
+          new Operator(this.context, this.scriptId, lastOutputType, null, this.onChildrenEvent())
+        )
```
Comment (Tommytrg) [inline src/script.ts]:
```suggestion
        const operator: MirOperator | null = type ? getDefaultMirOperatorByType(type): null
        this.operators.push(
          new Operator(this.context, this.scriptId, lastOutputType, operator, this.onChildrenEvent())
        )
```

File: src/script.ts [hunk 16]
```diff
@@ -94,13 +99,18 @@ export class Script {
...
 
   public getOutputType(): OutputType {
-    return this.getLastOperator()
-      ? this.operators.reduce((acc, operator) => {
-          let outputType = operator.operatorInfo.outputType
-          return outputType === OutputType.Same ? acc : outputType
-        }, this.firstType)
-      : this.firstType
-  }
+    console.log('source get output type', this.sourceType)
```
Comment (Tommytrg) [inline src/script.ts]:
```suggestion
```

File: src/source.ts [hunk 17]
```diff
@@ -1,23 +1,30 @@
...
+  constructor(context: { cache: Cache; i18n: I18n }, source: MirSource, sourceType: Kind, eventEmitter: EventEmitter) {
     this.id = context.cache.insert(this).id
-    this.kind = source.kind || 'HTTP-GET'
-    this.url = source.url || ''
-    this.contentType = source.contentType || 'JSON API'
-    this.script = new Script(context, source.script, OutputType.String)
+    this.kind = sourceType
+    this.kindOptions = KIND_OPTIONS
+    this.contentTypeOptions = CONTENT_TYPE_OPTIONS
+    this.url = sourceType === Kind.RNG ? '' : source.url || ''
```
Comment (Tommytrg) [inline src/source.ts]:
```suggestion
    this.url = sourceType === Kind.RNG ? '' : (source.url || '')
```

File: src/source.ts [hunk 18]
```diff
@@ -49,10 +60,22 @@ export class Source {
...
+    }
+  }
+
+  public update(args: { kind: Kind; url: string; contentType: string }) {
     const { kind = this.kind, url = this.url, contentType = this.contentType } = args
     this.kind = kind
+    this.kindOptions = this.kindOptions
     this.url = url
     this.contentType = contentType
+    console.log('emit', this.kind)
```
Comment (Tommytrg) [inline src/source.ts]:
```suggestion
```

**Review by Tommytrg — CHANGES_REQUESTED** [review]

**Review by Tommytrg — APPROVED** [review]
LGTM!

**Pull request merged**